client/packages/core/src/queryTypes.ts

```diff
@@ -70,9 +180,36 @@ type Direction = 'asc' | 'desc';
 
 type Order = { [key: string]: Direction };
 
-type $Option<Fields extends string[]> = {
+type $Option<
+  Fields extends string[],
+  WhereFieldTypes extends {
+    [key: string]: any;
+  } = Record<any, BSUnknown>,
+> = {
+  $?: {
+    where?: WhereClause<WhereFieldTypes>;
+    order?: Order;
+    limit?: number;
+    last?: number;
+    first?: number;
+    offset?: number;
+    after?: Cursor;
+    before?: Cursor;
+    fields?: Fields;
+  };
+};
+
+// Typed version that supports dot notation for nested queries
+type $OptionWNest<
+  S extends IContainEntitiesAndLinks<any, any>,
+  K extends keyof S['entities'],
+  Fields extends string[],
+  WhereFieldTypes extends {
+    [key: string]: any;
+  } = Record<any, BSUnknown>,
+> = {
   $?: {
-    where?: WhereClause;
+    where?: WhereClause<WhereFieldTypes> & TypedFreeWhereClause<S, K>;
```

Comment:
[to-discuss] Why do we need both `WhereClause` and `TypedFreeWhereClause`? Maybe we can get away with just one?